module/VuFind/src/VuFind/ILS/Driver/Folio.php

```diff
@@ -0,0 +1,579 @@
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * @category VuFind
+ * @package  ILS_Drivers
+ * @author   Chris Hallberg <challber@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:ils_drivers Wiki
+ */
+namespace VuFind\ILS\Driver;
+
+use VuFind\Exception\BadRequest as BadRequest;
+use VuFind\Exception\Forbidden as Forbidden;
+use VuFind\Exception\ILS as ILSException;
+use VuFind\Exception\RecordMissing as RecordMissing;
+use VuFind\I18n\Translator\TranslatorAwareInterface;
+
+/**
+ * FOLIO REST API driver
+ *
+ * @category VuFind
+ * @package  ILS_Drivers
+ * @author   Chris Hallberg <challber@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:ils_drivers Wiki
+ */
+class Folio extends AbstractAPI implements TranslatorAwareInterface
+{
+    use \VuFind\I18n\Translator\TranslatorAwareTrait;
+
+    /**
+     * Authentication token (X-Okapi-Token)
+     *
+     * @var string
+     */
+    protected $token = null;
+
+    /**
+     * Factory function for constructing the SessionContainer.
+     *
+     * @var Callable
+     */
+    protected $sessionFactory;
+
+    /**
+     * Session cache
+     *
+     * @var \Zend\Session\Container
+     */
+    protected $sessionCache;
+
+    /**
+     * Constructor
+     *
+     * @param \VuFind\Date\Converter $dateConverter  Date converter object
+     * @param Callable               $sessionFactory Factory function returning
+     * SessionContainer object
+     */
+    public function __construct(\VuFind\Date\Converter $dateConverter,
+        $sessionFactory
+    ) {
+        $this->dateConverter = $dateConverter;
+        $this->sessionFactory = $sessionFactory;
+    }
+
+    /**
+     * (From AbstractAPI) Allow default corrections to all requests
+     *
+     * Add X-Okapi headers and Content-Type to every request
+     *
+     * @param \Zend\Http\Headers $headers the request headers
+     * @param object             $params  the parameters object
+     *
+     * @return array
+     */
+    protected function preRequest(\Zend\Http\Headers $headers, $params)
+    {
+        $headers->addHeaderLine('Accept', 'application/json');
+        $headers->addHeaderLine('X-Okapi-Tenant', $this->config['API']['tenant']);
+        if (!$headers->has('Content-Type')) {
+            $headers->addHeaderLine('Content-Type', 'application/json');
+        }
+        if ($this->token != null) {
+            error_log('add token');
+            $headers->addHeaderLine('X-Okapi-Token', $this->token);
+        }
+        return [$headers, $params];
+    }
+
+    /**
+     * Login and receive a new token
+     *
+     * @return void
+     */
+    protected function renewTenantToken()
+    {
+        error_log('renewTenantToken');
```

Comment:
Since you're logger-aware, you should be able to use the built-in $this->debug instead of error_log; that might be preferable, at least for any debugging that you plan to keep in the long term. You can use the [Logging] section of config.ini to capture debug output to a separate log file if that would be helpful.